### PR TITLE
Add transaction lock when migrating configs

### PR DESCRIPTION
**Pull request by harshavardhana**
## Description
Add transaction lock when migrating configs

## Motivation and Context
When migrating configs it happens often that some
servers fail to start due to version mismatch etc.

Hold a transaction lock such that all servers get
serialized.

## Regression
Not sure

## How Has This Been Tested?
Just test this PR with or without this change. Without this change, some servers will fail to start from the following script.

```sh
#!/bin/bash

export MINIO_ERASURE_SET_DRIVE_COUNT=6
export MINIO_ACCESS_KEY="minio"
export MINIO_SECRET_KEY="minio123"

rm -rf /tmp/export{01..24}

for i in {01..24}; do
    minio server --address ":90${i}" http://127.0.0.1:9001/tmp/export01 http://127.0.0.1:9002/tmp/export02 http://127.0.0.1:9003/tmp/export03 http://127.0.0.1:9004/tmp/export04 http://127.0.0.1:9005/tmp/export05 http://127.0.0.1:9006/tmp/export06 http://127.0.0.1:9007/tmp/export07 http://127.0.0.1:9008/tmp/export08 http://127.0.0.1:9009/tmp/export09 http://127.0.0.1:9010/tmp/export10 http://127.0.0.1:9011/tmp/export11 http://127.0.0.1:9012/tmp/export12 http://127.0.0.1:9013/tmp/export13 http://127.0.0.1:9014/tmp/export14 http://127.0.0.1:9015/tmp/export15 http://127.0.0.1:9016/tmp/export16 http://127.0.0.1:9017/tmp/export17 http://127.0.0.1:9018/tmp/export18 http://127.0.0.1:9019/tmp/export19 http://127.0.0.1:9020/tmp/export20 http://127.0.0.1:9021/tmp/export21 http://127.0.0.1:9022/tmp/export22 http://127.0.0.1:9023/tmp/export23 http://127.0.0.1:9024/tmp/export24 &
done
```

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added unit tests to cover my changes.
- [ ] I have added/updated functional tests in [mint](https://github.com/minio/mint). (If yes, add `mint` PR # here: )
- [x] All new and existing tests passed.

### CODE REVIEW

**Comment by minio-ops** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-tls.sh | :heavy_check_mark: |
| mint-compression-xl.sh | :heavy_check_mark: |
| mint-xl.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-compression-fs.sh | :heavy_check_mark: |
| mint-worm.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-dist-xl.sh | :heavy_check_mark: |

**Review by krisis — APPROVED** [review]

File: cmd/config-migrate.go [hunk 1]
```diff
@@ -2413,8 +2413,58 @@ func migrateV27ToV28() error {
 	return nil
 }
 
+// Migrates ${HOME}/.minio/config.json to '<export_path>/.minio.sys/config/config.json'
+func migrateConfigToMinioSys(objAPI ObjectLayer) error {
+	defer os.Rename(getConfigFile(), getConfigFile()+".deprecated")
```
Comment (krishnasrinivas) [inline cmd/config-migrate.go]: Should be renamed only if:
1) checkConfig returns nil
OR
2) saveServerConfig() returns nil

File: cmd/config-migrate.go [hunk 2]
```diff
@@ -2413,8 +2413,58 @@ func migrateV27ToV28() error {
...
+
+	configFile := path.Join(minioConfigPrefix, minioConfigFile)
+	// Construct path to config.json for the given bucket.
+	transactionConfigFile := configFile + ".transaction"
```
Comment (krishnasrinivas) [inline cmd/config-migrate.go:2428]: lock can be held directly on `configFile` instead of `transactionConfigFile`

Comment (harshavardhana) [reply cmd/config-migrate.go:2428]: As discussed will not change this

File: cmd/config-migrate.go (hunk 1 above)
Comment (harshavardhana) [reply cmd/config-migrate.go]: Sure this can be done

**Review by krishnasrinivas — APPROVED** [review]

**Comment by minio-ops** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-tls.sh | :heavy_check_mark: |
| mint-compression-xl.sh | :heavy_check_mark: |
| mint-xl.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-compression-fs.sh | :heavy_check_mark: |
| mint-worm.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-dist-xl.sh | :x: [more...](#a2667d51-98fa-44fb-a91a-1d4acc47e6aa) |
<a name="a2667d51-98fa-44fb-a91a-1d4acc47e6aa"></a>
### 6878-2043f79/mint-dist-xl.sh.log:
```
Running with
SERVER_ENDPOINT:      minikube:32654
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp c37793c9d5ed:/mint/log /tmp/mint-logs'

(1/13) Running awscli tests ... done in 2 minutes and 23 seconds
(2/13) Running aws-sdk-go tests ... done in 1 seconds
(3/13) Running aws-sdk-java tests ... done in 5 seconds
(4/13) Running aws-sdk-php tests ... done in 46 seconds
(5/13) Running aws-sdk-ruby tests ... done in 18 seconds
(6/13) Running mc tests ... done in 1 minutes and 6 seconds
(7/13) Running minio-dotnet tests ... FAILED in 5 minutes and 35 seconds

Unhandled Exception: System.AggregateException: One or more errors occurred. (One or more errors occurred. (Minio API responded with message=We encountered an internal error, please try again.) (Minio API responded with message=We encountered an internal error, please try again.)) ---> System.AggregateException: One or more errors occurred. (Minio API responded with message=We encountered an internal error, please try again.) (Minio API responded with message=We encountered an internal error, please try again.) ---> Minio.Exceptions.MinioException: Minio API responded with message=We encountered an internal error, please try again.
   at Minio.MinioClient.ParseError(IRestResponse response) in /q/.q/sources/minio-dotnet/Minio/MinioClient.cs:line 363
   at Minio.MinioClient.HandleIfErrorResponse(IRestResponse response, IEnumerable`1 handlers, DateTime startTime) in /q/.q/sources/minio-dotnet/Minio/MinioClient.cs:line 505
   at Minio.MinioClient.ExecuteTaskAsync(IEnumerable`1 errorHandlers, IRestRequest request, CancellationToken cancellationToken) in /q/.q/sources/minio-dotnet/Minio/MinioClient.cs:line 351
   at Minio.MinioClient.PutObjectAsync(String bucketName, String objectName, String uploadId, Int32 partNumber, Byte[] data, Dictionary`2 metaData, CancellationToken cancellationToken) in /q/.q/sources/minio-dotnet/Minio/ApiEndpoints/ObjectOperations.cs:line 483
   at Minio.MinioClient.PutObjectAsync(String bucketName, String objectName, Stream data, Int64 size, String contentType, Dictionary`2 metaData, CancellationToken cancellationToken) in /q/.q/sources/minio-dotnet/Minio/ApiEndpoints/ObjectOperations.cs:line 229
   at Minio.Functional.Tests.FunctionalTest.PutObject_Task(MinioClient minio, String bucketName, String objectName, String fileName, String contentType, Int64 size, Dictionary`2 metaData, MemoryStream mstream) in /mint/run/core/minio-dotnet/FunctionalTest.cs:line 740
   --- End of inner exception stack trace ---
   at System.Threading.Tasks.Task.Wait(Int32 millisecondsTimeout, CancellationToken cancellationToken)
   at System.Threading.Tasks.Task.Wait()
   at Minio.Functional.Tests.FunctionalTest.RemoveObjects_Test2(MinioClient minio) in /mint/run/core/minio-dotnet/FunctionalTest.cs:line 1738
   --- End of inner exception stack trace ---
   at System.Threading.Tasks.Task.Wait(Int32 millisecondsTimeout, CancellationToken cancellationToken)
   at System.Threading.Tasks.Task.Wait()
   at Minio.Functional.Tests.FunctionalTest.Main(String[] args) in /mint/run/core/minio-dotnet/FunctionalTest.cs:line 218

Executed 6 out of 13 tests successfully.
```

**Pull request merged**